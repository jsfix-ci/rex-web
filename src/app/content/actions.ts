import {createStandardAction} from 'typesafe-actions';

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
