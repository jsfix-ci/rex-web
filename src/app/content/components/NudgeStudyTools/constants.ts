export const spotlightPadding = 1;
