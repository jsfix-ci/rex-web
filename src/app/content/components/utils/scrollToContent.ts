import scrollTo from 'scroll-to-element';
import { mobileBreak } from '../../../theme';
import {
  bookBannerDesktopMiniHeight,
  bookBannerMobileMiniHeight,
  toolbarDesktopHeight,
  toolbarMobileHeight
} from '../constants';

const pxPerRem = 10;
// height of navbar in px
const desktopOffset = -pxPerRem * (bookBannerDesktopMiniHeight + toolbarDesktopHeight);
const mobileOffset = -pxPerRem * (bookBannerMobileMiniHeight + toolbarMobileHeight);

export default function scrollToContent(elem: Element | string) {
    // innerWidth works best on Chrome
    // check other browsers at https://ryanve.com/lab/dimensions/
    // default to desktopOffset if window is undefined (at least the target is always visible)
    const offset = window && window.innerWidth <= pxPerRem * mobileBreak ? mobileOffset : desktopOffset;
    return scrollTo(elem, {offset});
}
