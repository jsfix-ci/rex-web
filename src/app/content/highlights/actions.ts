import { NewHighlight, UpdateHighlightRequest, HighlightColorEnum } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { HighlightData, SummaryHighlights } from './types';

export const focusHighlight = createStandardAction('Content/Highlight/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlight/clear')();
export const createHighlight = createStandardAction('Content/Highlight/create')<NewHighlight & {id: string}>();
export const deleteHighlight = createStandardAction('Content/Highlight/delete')<string>();
export const updateHighlight = createStandardAction('Content/Highlight/update')<UpdateHighlightRequest>();
export const receiveHighlights = createStandardAction('Content/Highlight/receive')<HighlightData[]>();

export const openMyHighlights = createStandardAction('Content/openMyHighlights')<void>();
export const closeMyHighlights = createStandardAction('Content/closeMyHighlights')<void>();

export const setColorsFilter = createStandardAction('Content/setColorsFilter')<HighlightColorEnum[]>();
export const setChaptersFilter = createStandardAction('Content/setChaptersFilter')<string[]>();
export const filtersChange = createStandardAction('Content/filtersChange')<void>();
export const setIsLoadingSummary = createStandardAction('Content/setIsLoadingSummary')<boolean>();
export const receiveSummaryHighlights = createStandardAction('Content/receiveSummaryHighlights')<SummaryHighlights>();
