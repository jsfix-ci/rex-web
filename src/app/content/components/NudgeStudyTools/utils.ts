import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { toolbarButtonMargin } from '../constants';
import {
  arrowDesktopHeight,
  arrowLeftMargin,
  arrowMobileHeight,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  contentMarginTop,
  cookieNudgeStudyGuidesCounter,
  cookieNudgeStudyGuidesDate,
  cookieNudgeStudyGuidesPageCounter,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  spotlightPadding,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperRight: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const target = useGetStudyToolsTarget();

  React.useEffect(() => {
    if (target) {
      const { top, left, right, height, width } = target.getBoundingClientRect();
      const padding = remsToPx(spotlightPadding);
      const spotlightTopOffset = top - padding;
      const spotlightLeftOffset = left - padding;
      const spotlightHeight = height + (padding * 2);
      const spotlightWidth = width + (padding * 2) - (isMobile ? 0 : remsToPx(toolbarButtonMargin));
      const arrowTopOffset = spotlightTopOffset + spotlightHeight + remsToPx(arrowTopMargin);
      const arrowLeft = spotlightLeftOffset + remsToPx(arrowLeftMargin);
      const contentWrapperTopOffset = arrowTopOffset
        + remsToPx(isMobile ? arrowMobileHeight : arrowDesktopHeight)
        + remsToPx(contentMarginTop);
      const contentWrapperRight = windowWidth - right - padding;
      const closeButtonLeft = left + spotlightWidth;
      const closeButtonTopOffset = contentWrapperTopOffset - remsToPx(closeButtonDistanceFromContent);

      setPositions({
        arrowLeft,
        arrowTopOffset,
        closeButtonLeft,
        closeButtonTopOffset,
        contentWrapperRight,
        contentWrapperTopOffset,
        spotlightHeight,
        spotlightLeftOffset,
        spotlightTopOffset,
        spotlightWidth,
      });
    }
    return () => setPositions(null);
  }, [target, windowWidth, isMobile]);

  return positions;
};

const useGetStudyToolsTarget = () => {
  const document = assertDocument();
  const [target, setTarget] = React.useState<HTMLElement | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    if (studyGuides) {
      setTarget(document.querySelector(`#${nudgeStudyToolsTargetId}`) as HTMLElement | null);
    }
    return () => setTarget(null);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [studyGuides]);

  return target;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudgeStudyGuidesCounter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudgeStudyGuidesDate);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const now = new Date();
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (now.getTime() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudgeStudyGuidesPageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  Cookies.set(cookieNudgeStudyGuidesPageCounter, (counter + 1).toString());
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudgeStudyGuidesCounter, (counter + 1).toString());
  Cookies.set(cookieNudgeStudyGuidesDate, now.toString());
  Cookies.remove(cookieNudgeStudyGuidesPageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  return React.useEffect(() => {
    const counter = getPageCounterCookie();
    if (page && counter <= nudgeStudyToolsMinPageLimit) { incrementPageCounterCookie(); }
  }, [page]);
};
