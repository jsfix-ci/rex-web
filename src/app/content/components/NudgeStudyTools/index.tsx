import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import ReactDOM from 'react-dom';
import { FormattedMessage } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import { useDebouncedMatchMobileQuery } from '../../../reactUtils';
import { assertDocument, assertNotNull } from '../../../utils';
import { closeNudgeStudyTools, openNudgeStudyTools } from '../../actions';
import { showNudgeStudyTools } from '../../selectors';
import { studyGuidesSummaryIsNotEmpty } from '../../studyGuides/selectors';
import arrowDesktop from './assets/arrowDesktop.svg';
import arrowMobile from './assets/arrowMobile.svg';
import {
  cookieNudgeStudyGuidesCounter,
  cookieNudgeStudyGuidesDate,
  nudgeStudyToolsShowLimit,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';
import {
  NudgeArrow,
  NudgeBackground,
  NudgeCloseButton,
  NudgeCloseIcon,
  NudgeContent,
  NudgeContentWrapper,
  NudgeHeading,
  NudgeSpotlight,
  NudgeText,
  NudgeWrapper,
} from './styles';
import { useGetStudyToolsTarget, usePositions } from './utils';

// tslint:disable-next-line: variable-name
const NudgeStudyTools = () => {
  const document = assertDocument();
  const body = assertNotNull(document.querySelector('body'), 'body element is not defined');
  const isMobile = useDebouncedMatchMobileQuery();
  const show = useSelector(showNudgeStudyTools);
  const target = useGetStudyToolsTarget();
  const positions = usePositions(target, isMobile);
  const hasStudyGuides = useSelector(studyGuidesSummaryIsNotEmpty);
  const trackOpen = useAnalyticsEvent('openNudgeStudyTools');
  const wrapperRef = React.useRef<HTMLElement>(null);
  const dispatch = useDispatch();

  React.useEffect(() => {
    if (show === null && hasStudyGuides) {
      const counter = Number(Cookies.get(cookieNudgeStudyGuidesCounter) || 0);
      const lastShownDate = Cookies.get(cookieNudgeStudyGuidesDate);
      const now = new Date();
      const passedAllowedInterval = Boolean(lastShownDate
        && (now.getTime() - new Date(lastShownDate).getTime()) > timeIntervalBetweenShowingNudgeInMs);

      if (
        counter < nudgeStudyToolsShowLimit
        && (!lastShownDate || passedAllowedInterval)
      ) {
        Cookies.set(cookieNudgeStudyGuidesCounter, (counter + 1).toString());
        Cookies.set(cookieNudgeStudyGuidesDate, now.toString());
        trackOpen();
        dispatch(openNudgeStudyTools());
      }
    }
  }, [show, hasStudyGuides, trackOpen, dispatch]);

  React.useEffect(() => {
    if (show && positions) {
      body.style.overflow = 'hidden';
    }
    return () => { body.style.overflow = null; };
  }, [body, show, positions]);

  React.useEffect(() => {
    if (show && positions && wrapperRef.current) {
      wrapperRef.current.focus();
    }
  }, [show, positions, wrapperRef]);

  if (!show || !target || !positions) { return null; }

  const close = () => {
    dispatch(closeNudgeStudyTools());
  };

  return ReactDOM.createPortal(<FormattedMessage id='i18n:nudge:study-tools:aria-label'>
    {(msg: string) => <NudgeWrapper
      ref={wrapperRef}
      tabIndex='-1'
      aria-label={msg}
      data-analytics-region='Nudge Study Tools'
    >
      <NudgeArrow
        src={isMobile ? arrowMobile : arrowDesktop}
        alt=''
        top={positions.arrowTopOffset}
        left={positions.arrowLeft}
      />
      <NudgeCloseButton
        top={positions.closeButtonTopOffset}
        left={positions.closeButtonLeft}
        onClick={close}
        data-analytics-label='close'
      >
        <NudgeCloseIcon />
      </NudgeCloseButton>
      <NudgeContentWrapper
        top={positions.contentWrapperTopOffset}
        right={positions.contentWrapperRight}
      >
        <NudgeContent>
          <NudgeHeading />
          <NudgeText />
        </NudgeContent>
      </NudgeContentWrapper>
      <NudgeBackground>
        <NudgeSpotlight
          top={positions.spotlightTopOffset}
          left={positions.spotlightLeftOffset}
          height={positions.spotlightHeight}
          width={positions.spotlightWidth}
        />
      </NudgeBackground>
    </NudgeWrapper>}
  </FormattedMessage>,
  body
  );
};

export default NudgeStudyTools;
