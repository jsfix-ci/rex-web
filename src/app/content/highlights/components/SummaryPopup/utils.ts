import { Highlight } from '@openstax/highlighter/dist/api';
import { generateTargetQueryFromScrollTarget } from '../../../../navigation/utils';
import { Book } from '../../../types';
import { findArchiveTreeNode } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const createHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const page = book ? findArchiveTreeNode(book.tree, sourceId) : undefined;
  const target = generateTargetQueryFromScrollTarget({ id, type: 'highlight', elementId });
  return `${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`;
};
