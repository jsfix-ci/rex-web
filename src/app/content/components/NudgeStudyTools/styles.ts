import styled, { css } from 'styled-components/macro';
import { Times } from 'styled-icons/fa-solid/Times/Times';
import { PlainButton } from '../../../components/Button';
import htmlMessage from '../../../components/htmlMessage';
import theme from '../../../theme';
import { remsToPx } from '../../../utils';
import { arrowDesktopHeight, arrowMobileHeight, closeButtonMobileMargin } from './constants';

// tslint:disable-next-line: variable-name
export const NudgeWrapper = styled.div`
  display: contents;
`;

// tslint:disable-next-line: variable-name
export const NudgeContentWrapper = styled.div`
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  position: absolute;
  width: max-content;
  top: 100%;
  ${theme.breakpoints.mobile(css`
    right: auto;
    text-align: center;
    padding: 0 2rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeContent = styled.div`
  position: relative;
`;

export const Container = styled.div`
  display:flex;
  flex-direction:column;
  overflow:visible;
  /*just needs to be higher than the cover*/
  z-index: 9999;
  position: relative;
`

export const FittingSizeContainer = styled.div`
  width: 100%;
`

export const Spotlight = styled.div`
  background: white;
  display: flex;
  padding: 10px;
`

export const InnerContainer = styled.div`
  display:flex;
  position: absolute;
  flex-direction:column;
  align-items:flex-end;
  width: 100%;
  overflow: visible;
  top: 100%;
`

// tslint:disable-next-line: variable-name
const NudgeHeadingStyles = styled.h2`
  font-size: 3.6rem;
  line-height: 1.1;
  letter-spacing: -1.4px;
  margin: 0 0 1.7rem 0;
  color: ${theme.color.text.white};
  overflow: hidden;

  strong {
    color: ${theme.color.primary.yellow.base};
    font-weight: 600;
  }

  ${theme.breakpoints.mobile(css`
    font-size: 2.4rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeHeading = htmlMessage('i18n:nudge:study-tools:heading', NudgeHeadingStyles);

// tslint:disable-next-line: variable-name
const NudgeTextStyles = styled.div`
  font-size: 2.4rem;
  font-weight: 400;
  line-height: 1.1;
  letter-spacing: -1px;
  color: ${theme.color.text.white};
  max-width: 690px;
  overflow: hidden;
  ${theme.breakpoints.mobile(css`
    max-width: 100%;
    font-size: 1.6rem;
    line-height: 1.4;
    padding: 0 2rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeText = htmlMessage('i18n:nudge:study-tools:text', NudgeTextStyles);

// tslint:disable-next-line: variable-name
export const NudgeArrow = styled.img`
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  display: block;
  height: ${arrowDesktopHeight}rem;

  ${theme.breakpoints.mobile(css`
    height: ${arrowMobileHeight}rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeCloseIcon = styled(Times)`
  width: 1.1rem;
  color: ${theme.color.text.white};
`;

// tslint:disable-next-line: variable-name
export const NudgeCloseButton = styled(PlainButton)`
  z-index: ${theme.zIndex.nudgeOverlay + 1};
  width: 4rem;
  height: 4rem;
  padding: 1rem;
  border-radius: 50%;
  border: 1px solid ${theme.color.white};
  display: flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  right: 0;
  transform:translateY(-100%);
  ${theme.breakpoints.mobile(css`
    left: auto;
    top: ${remsToPx(closeButtonMobileMargin)}px;
    right: ${remsToPx(closeButtonMobileMargin)}px;
  `)}
`;

// tslint:disable-next-line: variable-name
export const NudgeBackground = styled.div`
  position: absolute;
  width: 200vw;
  height: 200vh;
  left: 0%;
  z-index: ${theme.zIndex.nudgeOverlay + 2};
  inset: 0px;
  opacity: 0.9;
  background-color: ${theme.color.black};
`;

// tslint:disable-next-line: variable-name
export const NudgeSpotlight = styled.div`
  position: fixed;
  background-color: gray;
`;
