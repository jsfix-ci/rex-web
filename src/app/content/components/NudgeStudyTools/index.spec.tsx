import { Document } from '@openstax/types/lib.dom';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { AppServices, Store } from '../../../types';
import { assertDocument } from '../../../utils';
import { openNudgeStudyTools } from '../../actions';
import * as contentSelect from '../../selectors';
import * as studyGuidesSelect from '../../studyGuides/selectors';
import NudgeStudyTools from './';
import * as utils from './utils';

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: (component: any) => component,
}));

describe('NudgeStudyTools', () => {
  let document: Document;
  let store: Store;
  let dispatch: jest.SpyInstance;
  let services: AppServices;

  beforeEach(() => {
    document = assertDocument();

    store = createTestStore();
    dispatch = jest.spyOn(store, 'dispatch');
    services = createTestServices();
  });

  it('sets cookies, opens nudge and track opening if all requirement passes', () => {
    jest.spyOn(studyGuidesSelect, 'hasStudyGuides')
      .mockReturnValue(true);
    jest.spyOn(utils, 'shouldDisplayNudgeStudyTools')
      .mockReturnValue(true);
    const spySetCookies = jest.spyOn(utils, 'setNudgeStudyToolsCookies');
    const spyTrack = jest.spyOn(services.analytics.openNudgeStudyTools, 'track');

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <NudgeStudyTools/>
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    // Call useEffect hooks
    component.update(() => null);

    expect(spySetCookies).toHaveBeenCalled();
    expect(spyTrack).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(openNudgeStudyTools());
  });

  it('does not render if not open', () => {
    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <NudgeStudyTools/>
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('does not render if open but without positions', () => {
    jest.spyOn(contentSelect, 'showNudgeStudyTools')
      .mockReturnValue(true);

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <NudgeStudyTools/>
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders only if open and if positions have been calculated', async() => {
    jest.spyOn(contentSelect, 'showNudgeStudyTools')
      .mockReturnValue(true);

    jest.spyOn(utils, 'usePositions')
      .mockReturnValue({
        arrowLeft: 1200,
        arrowTopOffset: 245,
        closeButtonLeft: 1500,
        closeButtonTopOffset: 345,
        contentWrapperRight: -486,
        contentWrapperTopOffset: 385,
        spotlightHeight: 45,
        spotlightLeftOffset: 1190,
        spotlightTopOffset: 190,
        spotlightWidth: 300,
      });

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <NudgeStudyTools/>
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
