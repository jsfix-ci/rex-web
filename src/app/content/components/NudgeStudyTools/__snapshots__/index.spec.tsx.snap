// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NudgeStudyTools does not render if not open 1`] = `null`;

exports[`NudgeStudyTools does not render if open but without positions 1`] = `null`;

exports[`NudgeStudyTools renders only if open and if positions have been calculated 1`] = `
<div
  aria-label="Image points to the Study Guides and My Highlights buttons in the toolbar. The study tools you need. 100% free. Boost your studies with expert-created study guides and highlighting. To close image, navigate to or click the x button."
  className="styles__NudgeWrapper-hrv0cf-0 fJZGzd"
  data-analytics-region="Nudge Study Tools"
  tabIndex="-1"
>
  <img
    alt=""
    className="styles__NudgeArrow-hrv0cf-5 dAnHcf"
    src="arrowDesktop.svg"
  />
  <button
    className="Button__PlainButton-ayg7nk-2 styles__NudgeCloseButton-hrv0cf-7 kRuXQn"
    data-analytics-label="close"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="styles__NudgeCloseIcon-hrv0cf-6 gilaaP StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 352 512"
    >
      <path
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
  <div
    className="styles__NudgeContentWrapper-hrv0cf-1 dAyyBL"
  >
    <div
      className="styles__NudgeContent-hrv0cf-2 jYfSzb"
    >
      <h2
        className="styles__NudgeHeadingStyles-hrv0cf-3 ekEeMe"
        dangerouslySetInnerHTML={
          Object {
            "__html": "The study tools you need. <strong>100% FREE!</strong>",
          }
        }
      />
      <div
        className="styles__NudgeTextStyles-hrv0cf-4 jPNjeY"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Boost your studies with expert-created study guides and highlighting. <strong>It's all free.</strong>",
          }
        }
      />
    </div>
  </div>
  <div
    className="styles__NudgeBackground-hrv0cf-8 grqcfE"
  >
    <div
      className="styles__NudgeSpotlight-hrv0cf-9 bOQXlU"
      height={45}
      width={300}
    />
  </div>
</div>
`;
